### PR TITLE
#5 Videos API

### PR DESCRIPTION
**Pull request by JonaszJestem**

### CODE REVIEW

File: src/core/video/video.service.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          resolve(fs.promises.readFile('test.mp4'));
+        })
+        .on('error', (err) => {
+          console.log('an error happened: ' + err.message);
+          reject(err);
+        })
+        .size(`${options.height}x${options.width}`)
+        .format('mp4')
+        // TODO: Pass stream to fastify response somehow, do not create files
+        .saveToFile('test.mp4');
```
Comment (JonaszJestem) [inline src/core/video/video.service.ts:27]: I couldn't get this working with streams today. It works this way, but I will try to stream that tomorrow.

File: src/core/video/video.service.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        .input(file)
+        .on('end', () => {
+          resolve(true);
+        })
+        .on('error', (err) => {
+          reject(err);
+        })
+        .size(`${options.height}x${options.width}`)
+        .format('h264')
+        .pipe(undefined, {end: true});
```
Comment (JonaszJestem) [inline src/core/video/video.service.ts:25]: that doesn't work, response never goes back to client

File: src/core/video/video.service.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+          reject(err);
+        }
+
+        resolve(data);
+      });
+    });
+  }
+
+  public async thumbnail(file: Readable, options: { second: number }): Promise<Stream> {
+    const tempFileName = `${await createTempFilename()}.png`;
```
Comment (JonaszJestem) [inline src/core/video/video.service.ts:42]: According to docs: https://github.com/fluent-ffmpeg/node-fluent-ffmpeg#screenshotsoptions-dirname-generate-thumbnails
it doesn't work on input streams. That's why I used temporary file - it should be small enough to work pretty fast.

File: src/core/video/video.service.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  public async metadata(file: Readable): Promise<FfprobeData> {
+    return await new Promise((resolve, reject) => {
+      ffmpeg().input(file).ffprobe((err, data) => {
+        if (err) {
+          reject(err);
+        }
+
+        resolve(data);
```
Comment (JonaszJestem) [inline src/core/video/video.service.ts:36]: Do we want to serialize the response? Currently it's raw output which looks like this:
```
{
  "streams": [
    {
      "index": 0,
      "codec_name": "h264",
      "codec_long_name": "H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10",
      "profile": "High",
      "codec_type": "video",
      "codec_tag_string": "avc1",
      "codec_tag": "0x31637661",
      "width": 1920,
      "height": 1080,
      "coded_width": 1920,
      "coded_height": 1080,
      "closed_captions": 0,
      "has_b_frames": 2,
      "sample_aspect_ratio": "1:1",
      "display_aspect_ratio": "16:9",
      "pix_fmt": "yuv420p",
      "level": 40,
      "color_range": "unknown",
      "color_space": "unknown",
      "color_transfer": "unknown",
      "color_primaries": "unknown",
      "chroma_location": "left",
      "field_order": "unknown",
      "refs": 1,
      "is_avc": "true",
      "nal_length_size": 4,
      "id": "N/A",
      "r_frame_rate": "30000/1001",
      "avg_frame_rate": "30000/1001",
      "time_base": "1/30000",
      "start_pts": 0,
      "start_time": 0,
      "duration_ts": 100100,
      "duration": 3.336667,
      "bit_rate": 763463,
      "max_bit_rate": "N/A",
      "bits_per_raw_sample": 8,
      "nb_frames": 100,
      "nb_read_frames": "N/A",
      "nb_read_packets": "N/A",
      "tags": {
        "language": "und",
        "handler_name": "Apple Video Media Handler",
        "vendor_id": "[0][0][0][0]"
      },
      "disposition": {
        "default": 1,
        "dub": 0,
        "original": 0,
        "comment": 0,
        "lyrics": 0,
        "karaoke": 0,
        "forced": 0,
        "hearing_impaired": 0,
        "visual_impaired": 0,
        "clean_effects": 0,
        "attached_pic": 0,
        "timed_thumbnails": 0
      }
    }
  ],
  "format": {
    "filename": "pipe:0",
    "nb_streams": 1,
    "nb_programs": 0,
    "format_name": "mov,mp4,m4a,3gp,3g2,mj2",
    "format_long_name": "QuickTime / MOV",
    "start_time": 0,
    "duration": 3.337,
    "size": "N/A",
    "bit_rate": "N/A",
    "probe_score": 100,
    "tags": {
      "major_brand": "isom",
      "minor_version": "512",
      "compatible_brands": "isomiso2avc1mp41",
      "encoder": "Lavf58.45.100"
    }
  },
  "chapters": []
}
```

File: src/core/video/video.service.ts (hunk 3 above)
Comment (JonaszJestem) [reply src/core/video/video.service.ts:42]: Or maybe you have some examples on how did you resolve that in other project?

File: src/core/video/video.service.ts (hunk 4 above)
Comment (JonaszJestem) [reply src/core/video/video.service.ts:36]: nevermind, I found the issue with details: https://github.com/leocode/image-service/issues/2

File: src/core/video/video.service.ts (hunk 3 above)
Comment (soanvig) [reply src/core/video/video.service.ts:42]: That's ok imho

**Review by soanvig — APPROVED** [review]

File: src/core/video/handlers/resize.handler.ts [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+import { VideoService } from '../video.service';
+
+type ResizeQuery = { width: number; height: number, codecName: string };
+
+const resizeQuerySchema = {
+  type: 'object',
+  properties: {
+    height: { type: 'number' },
+    width: { type: 'number' },
+    codecName: { type: 'string' },
```
Comment (KrzysiekWyka) [inline src/core/video/handlers/resize.handler.ts:14]: @JonaszJestem Can u convince me this property is useful? From client's perspective, all I want to do is resize video, now I also have to know its codec.

Comment (JonaszJestem) [reply src/core/video/handlers/resize.handler.ts:14]: Resolved in #17

**Review by KrzysiekWyka — APPROVED** [review]

**Pull request merged**